spec/unit/Adb.spec.js

```diff
@@ -0,0 +1,222 @@
+        });
+    });
+
+    describe('devices', () => {
+        beforeEach(() => {
+            spawnSpy.and.returnValue(Promise.resolve(adbOutput));
+        });
+
+        it('should return only devices if no options are specified', () => {
+            return Adb.devices().then(devices => {
+                expect(devices.length).toBe(1);
+                expect(devices[0]).toBe(deviceId);
+            });
+        });
+
+        it('should return only emulators if opts.emulators is true', () => {
+            return Adb.devices({emulators: true}).then(devices => {
+                expect(devices.length).toBe(1);
+                expect(devices[0]).toBe(emulatorId);
+            });
+        });
+    });
+
+    describe('install', () => {
+        it('should target the passed device id to adb', () => {
+            spawnSpy.and.returnValue(Promise.resolve(''));
+
+            return Adb.install(deviceId).then(() => {
+                const args = spawnSpy.calls.argsFor(0);
+                expect(args[0]).toBe('adb');
+
+                const adbArgs = args[1].join(' ');
+                expect(adbArgs).toMatch(`-s ${deviceId}`);
+            });
+        });
+
+        it('should add the -r flag if opts.replace is set', () => {
+            spawnSpy.and.returnValue(Promise.resolve(''));
+
+            return Adb.install(deviceId, '', { replace: true }).then(() => {
+                const adbArgs = spawnSpy.calls.argsFor(0)[1];
+                expect(adbArgs).toContain('-r');
+            });
+        });
+
+        it('should pass the correct package path to adb', () => {
+            const packagePath = 'build/test/app.apk';
+            spawnSpy.and.returnValue(Promise.resolve(''));
+
+            return Adb.install(deviceId, packagePath).then(() => {
+                const adbArgs = spawnSpy.calls.argsFor(0)[1];
+                expect(adbArgs).toContain(packagePath);
+            });
+        });
+
+        it('should reject with a CordovaError if the adb output suggests a failure', () => {
+            spawnSpy.and.returnValue(Promise.resolve(alreadyExistsError));
+
+            return Adb.install(deviceId, '')
+                .then(() => fail('Unexpectedly resolved'))
+                .catch(err => {
+                    expect(err instanceof Adb.__get__('CordovaError')).toBe(true);
+                });
+        });
+
+        // The following two tests are somewhat brittle as they are dependent on the
+        // exact message returned. But it is better to have them tested than not at all.
+        it('should give a more specific error message if there is a certificate failure', () => {
+            spawnSpy.and.returnValue(Promise.resolve(certificateError));
+
+            return Adb.install(deviceId, '')
+                .then(() => fail('Unexpectedly resolved'))
+                .catch(err => {
+                    expect(err).toMatch('Sign the build');
+                });
```

Comment:
I asked because I would assert that err is an error whenever possible. Debugging thrown primitives can be really nasty, due to the lack of a stack trace.

So if we actually get a CordovaError here, I would add
```js
expect(err).toEqual(jasmine.any(CordovaError));
```